src/backend/distributed/worker/worker_data_fetch_protocol.c

```diff
@@ -526,30 +525,46 @@ worker_apply_sequence_command(PG_FUNCTION_ARGS)
 }
 
 
-/* Extracts shard id from the given table name, and returns it. */
-static uint64
-ExtractShardId(const char *tableName)
+/*
+ * ExtractShardIdFromTableName tries to extract shard id from the given table name,
+ * and returns the shard id if table name is formatted as shard name.
+ * Else, the function returns INVALID_SHARD_ID.
+ */
+uint64
+ExtractShardIdFromTableName(const char *tableName, bool missingOk)
 {
 	uint64 shardId = 0;
 	char *shardIdString = NULL;
 	char *shardIdStringEnd = NULL;
 
 	/* find the last underscore and increment for shardId string */
 	shardIdString = strrchr(tableName, SHARD_NAME_SEPARATOR);
-	if (shardIdString == NULL)
+	if (shardIdString == NULL && !missingOk)
 	{
 		ereport(ERROR, (errmsg("could not extract shardId from table name \"%s\"",
 							   tableName)));
 	}
+	else if (shardIdString == NULL && missingOk)
+	{
+		return INVALID_SHARD_ID;
+	}
+
 	shardIdString++;
 
 	errno = 0;
 	shardId = pg_strtouint64(shardIdString, &shardIdStringEnd, 0);
 
 	if (errno != 0 || (*shardIdStringEnd != '\0'))
 	{
-		ereport(ERROR, (errmsg("could not extract shardId from table name \"%s\"",
-							   tableName)));
+		if (!missingOk)
```

Comment:
Code repetition makes me feel uncomfortable here. Is there an easy way to avoid this ?

Reply:
Hmm, there are two steps involved in extracting the `shardId`.

1. Search for `_` in the table name
2. Try to convert `str` to `uint`

Two `missingOk`  seems should be added for each of the checks above. If you see a simplification, feel free to suggest, I can apply them